File: tests/bench.go
package main

import (
	"bufio"
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

type BenchResult struct {
	Label        string        `json:"label"`
	Model        string        `json:"model"`
	Stream       bool          `json:"stream"`
	PromptLen    int           `json:"prompt_tokens_est"`
	HTTPStatus   int           `json:"http_status"`
	TTFT         time.Duration `json:"ttft_ms"`
	TotalTime    time.Duration `json:"total_ms"`
	ChunkCount   int           `json:"chunk_count"`
	OutputTokens int           `json:"completion_tokens"`
	PromptTokens int           `json:"prompt_tokens"`
	ErrorBody    string        `json:"error_body,omitempty"`
}

type ChatRequest struct {
	Model     string        `json:"model"`
	Messages  []ChatMessage `json:"messages"`
	Stream    bool          `json:"stream"`
	MaxTokens int           `json:"max_tokens"`
}

type ChatMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type SSEChunk struct {
	Choices []struct {
		Delta struct {
			Content string `json:"content"`
		} `json:"delta"`
	} `json:"choices"`
	Usage *struct {
		PromptTokens     int `json:"prompt_tokens"`
		CompletionTokens int `json:"completion_tokens"`
	} `json:"usage,omitempty"`
}

type NonStreamResp struct {
	Choices []struct {
		Message struct {
			Content string `json:"content"`
		} `json:"message"`
	} `json:"choices"`
	Usage struct {
		PromptTokens     int `json:"prompt_tokens"`
		CompletionTokens int `json:"completion_tokens"`
	} `json:"usage"`
}

func main() {
	endpoint := flag.String("url", "http://localhost:8789/v1/chat/completions", "proxy url")
	flag.Parse()

	cases := []struct {
		label     string
		model     string
		stream    bool
		prompt    string
		maxTokens int
	}{
		{"nonstream-short-gemini", "google/gemini-2.5-flash", false, "用一句话介绍 Go 语言", 80},
		{"nonstream-short-haiku", "anthropic/claude-3.5-haiku-20241022", false, "用一句话介绍 Go 语言", 80},
		{"stream-short-gemini", "google/gemini-2.5-flash", true, "用一句话介绍 Go 语言", 80},
		{"stream-short-haiku", "anthropic/claude-3.5-haiku-20241022", true, "用一句话介绍 Go 语言", 80},
		{"stream-short-sonnet", "anthropic/claude-sonnet-4", true, "用一句话介绍 Go 语言", 80},
		{"stream-medium-gemini", "google/gemini-2.5-flash", true, strings.Repeat("Go 是一门由 Google 开发的静态类型编译型语言。", 10) + "\n\n请简要总结上述内容。", 150},
		{"stream-medium-haiku", "anthropic/claude-3.5-haiku-20241022", true, strings.Repeat("Go 是一门由 Google 开发的静态类型编译型语言。", 10) + "\n\n请简要总结上述内容。", 150},
	}

	results := make([]BenchResult, 0, len(cases))
	for _, c := range cases {
		fmt.Printf("[RUN] %s ... ", c.label)
		r := runOne(*endpoint, c.label, c.model, c.stream, c.prompt, c.maxTokens)
		results = append(results, r)
		if r.HTTPStatus == 200 {
			fmt.Printf("OK ttft=%v total=%v chunks=%d tokens=%d/%d\n",
				r.TTFT.Round(time.Millisecond), r.TotalTime.Round(time.Millisecond),
				r.ChunkCount, r.PromptTokens, r.OutputTokens)
		} else {
			fmt.Printf("FAIL http=%d err=%.120s\n", r.HTTPStatus, r.ErrorBody)
		}
		time.Sleep(500 * time.Millisecond)
	}

	out, _ := json.MarshalIndent(results, "", "  ")
	fmt.Println("\n=== RESULTS JSON ===")
	fmt.Println(string(out))
}

func runOne(endpoint, label, model string, stream bool, prompt string, maxTokens int) BenchResult {
	reqBody := ChatRequest{
		Model:     model,
		Messages:  []ChatMessage{{Role: "user", Content: prompt}},
		Stream:    stream,
		MaxTokens: maxTokens,
	}
	payload, _ := json.Marshal(reqBody)

	result := BenchResult{
		Label:     label,
		Model:     model,
		Stream:    stream,
		PromptLen: len(prompt),
	}

	httpReq, _ := http.NewRequest(http.MethodPost, endpoint, bytes.NewReader(payload))
	httpReq.Header.Set("Content-Type", "application/json")

	t0 := time.Now()
	client := &http.Client{Timeout: 120 * time.Second}
	resp, err := client.Do(httpReq)
	if err != nil {
		result.ErrorBody = err.Error()
		result.TotalTime = time.Since(t0)
		return result
	}
	defer resp.Body.Close()
	result.HTTPStatus = resp.StatusCode

	if resp.StatusCode != 200 {
		body, _ := io.ReadAll(resp.Body)
		result.ErrorBody = string(body)
		result.TotalTime = time.Since(t0)
		return result
	}

	if !stream {
		body, _ := io.ReadAll(resp.Body)
		result.TotalTime = time.Since(t0)
		result.TTFT = result.TotalTime
		var r NonStreamResp
		if err := json.Unmarshal(body, &r); err == nil {
			result.PromptTokens = r.Usage.PromptTokens
			result.OutputTokens = r.Usage.CompletionTokens
		}
		return result
	}

	scanner := bufio.NewScanner(resp.Body)
	scanner.Buffer(make([]byte, 0, 64*1024), 1024*1024)

	firstContent := false
	for scanner.Scan() {
		line := scanner.Text()
		if !strings.HasPrefix(line, "data: ") {
			continue
		}
		data := strings.TrimPrefix(line, "data: ")
		if strings.TrimSpace(data) == "[DONE]" {
			break
		}
		result.ChunkCount++
		var ch SSEChunk
		if err := json.Unmarshal([]byte(data), &ch); err != nil {
			continue
		}
		if !firstContent {
			for _, c := range ch.Choices {
				if c.Delta.Content != "" {
					result.TTFT = time.Since(t0)
					firstContent = true
					break
				}
			}
		}
		if ch.Usage != nil {
			result.PromptTokens = ch.Usage.PromptTokens
			result.OutputTokens = ch.Usage.CompletionTokens
		}
	}
	result.TotalTime = time.Since(t0)
	return result
}

File: tests/bench.py
"""
FreeBuff2API performance benchmark.
Tests streaming TTFT (time to first token), total time, and throughput.
"""
import json
import time
import sys
import urllib.request
import urllib.error

BASE_URL = "http://localhost:8787"


def bench_non_stream(model, prompt, max_tokens=200, label=""):
    payload = {
        "model": model,
        "messages": [{"role": "user", "content": prompt}],
        "stream": False,
        "max_tokens": max_tokens,
    }
    data = json.dumps(payload).encode("utf-8")
    req = urllib.request.Request(
        f"{BASE_URL}/v1/chat/completions",
        data=data,
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    t0 = time.time()
    try:
        with urllib.request.urlopen(req, timeout=180) as resp:
            body = resp.read().decode("utf-8", errors="replace")
            total = time.time() - t0
            try:
                obj = json.loads(body)
                usage = obj.get("usage", {})
                content = ""
                choices = obj.get("choices", [])
                if choices:
                    content = choices[0].get("message", {}).get("content", "")
                return {
                    "ok": True,
                    "label": label,
                    "model": model,
                    "total_s": total,
                    "prompt_tokens": usage.get("prompt_tokens"),
                    "completion_tokens": usage.get("completion_tokens"),
                    "total_tokens": usage.get("total_tokens"),
                    "content_preview": content[:120],
                }
            except Exception as e:
                return {"ok": False, "label": label, "model": model, "error": f"json: {e}", "body": body[:500]}
    except urllib.error.HTTPError as e:
        body = e.read().decode("utf-8", errors="replace")
        return {"ok": False, "label": label, "model": model, "error": f"HTTP {e.code}", "body": body[:500]}
    except Exception as e:
        return {"ok": False, "label": label, "model": model, "error": str(e)}


def bench_stream(model, prompt, max_tokens=300, label=""):
    payload = {
        "model": model,
        "messages": [{"role": "user", "content": prompt}],
        "stream": True,
        "max_tokens": max_tokens,
    }
    data = json.dumps(payload).encode("utf-8")
    req = urllib.request.Request(
        f"{BASE_URL}/v1/chat/completions",
        data=data,
        headers={"Content-Type": "application/json", "Accept": "text/event-stream"},
        method="POST",
    )
    t0 = time.time()
    ttft = None
    chunks = 0
    content_chars = 0
    completion_tokens = None
    prompt_tokens = None
    first_chunk_time = None
    done = False
    try:
        with urllib.request.urlopen(req, timeout=180) as resp:
            for raw in resp:
                try:
                    line = raw.decode("utf-8", errors="replace").strip()
                except Exception:
                    continue
                if not line:
                    continue
                if first_chunk_time is None:
                    first_chunk_time = time.time() - t0
                if line == "data: [DONE]":
                    done = True
                    break
                if line.startswith("data:"):
                    chunks += 1
                    body = line[5:].strip()
                    if not body:
                        continue
                    try:
                        obj = json.loads(body)
                    except Exception:
                        continue
                    # capture usage if present
                    u = obj.get("usage")
                    if u:
                        prompt_tokens = u.get("prompt_tokens", prompt_tokens)
                        completion_tokens = u.get("completion_tokens", completion_tokens)
                    for ch in obj.get("choices", []):
                        delta = ch.get("delta") or {}
                        c = delta.get("content")
                        if c:
                            if ttft is None:
                                ttft = time.time() - t0
                            content_chars += len(c)
            total = time.time() - t0
            return {
                "ok": True,
                "label": label,
                "model": model,
                "ttft_s": ttft,
                "first_chunk_s": first_chunk_time,
                "total_s": total,
                "chunks": chunks,
                "content_chars": content_chars,
                "prompt_tokens": prompt_tokens,
                "completion_tokens": completion_tokens,
                "done_sentinel": done,
                "throughput_tok_s": (completion_tokens / total) if completion_tokens and total > 0 else None,
                "throughput_chars_s": (content_chars / total) if total > 0 else None,
            }
    except urllib.error.HTTPError as e:
        body = e.read().decode("utf-8", errors="replace")
        return {"ok": False, "label": label, "model": model, "error": f"HTTP {e.code}", "body": body[:500]}
    except Exception as e:
        return {"ok": False, "label": label, "model": model, "error": str(e)}


SHORT_PROMPT = "用一句话介绍你自己"
# ~200 tokens CJK-ish prompt
MEDIUM_PROMPT = (
    "请用中文写一段约150字的文本，介绍一下计算机科学中的‘操作系统’是什么，"
    "包括它的主要职责（进程管理、内存管理、文件系统、设备管理、用户接口等），"
    "以及常见的操作系统举例（Linux、Windows、macOS、Android、iOS 等）。"
    "语言风格请保持科普、通俗、准确，避免过度学术化，结尾用一句话总结操作系统的重要性。"
)


def main():
    results = []

    print("=== Non-stream tests ===", flush=True)
    for i in range(2):
        r = bench_non_stream("google/gemini-2.5-flash", SHORT_PROMPT, max_tokens=200,
                             label=f"nonstream#{i+1} gemini-flash short")
        print(json.dumps(r, ensure_ascii=False), flush=True)
        results.append(r)

    print("=== Stream tests: short prompt ===", flush=True)
    for model in ["google/gemini-2.5-flash", "anthropic/claude-3.5-haiku-20241022", "anthropic/claude-sonnet-4"]:
        r = bench_stream(model, SHORT_PROMPT, max_tokens=300, label=f"stream short {model}")
        print(json.dumps(r, ensure_ascii=False), flush=True)
        results.append(r)

    print("=== Stream tests: medium prompt ===", flush=True)
    for model in ["google/gemini-2.5-flash", "anthropic/claude-3.5-haiku-20241022"]:
        r = bench_stream(model, MEDIUM_PROMPT, max_tokens=400, label=f"stream medium {model}")
        print(json.dumps(r, ensure_ascii=False), flush=True)
        results.append(r)

    # Save raw JSON
    with open("tests/results.json", "w", encoding="utf-8") as f:
        json.dump(results, f, ensure_ascii=False, indent=2)
    print("Saved tests/results.json", flush=True)


if __name__ == "__main__":
    main()
